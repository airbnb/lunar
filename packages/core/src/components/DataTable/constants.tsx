import { WidthProperties } from './types';

export const SELECTION_OPTIONS = {
  ACTIVE: 'ACTIVE',
  DISABLED: 'DISABLED',
  INACTIVE: 'INACTIVE',
  HAS_ACTIVE_CHILD: 'HAS_ACTIVE_CHILD',
};

export const STATUS_OPTIONS = {
  ALERT: 'ALERT',
  WARNING: 'WARNING',
};

type HeightMap = {
  [key: string]: number;
};

export const HEIGHT_TO_PX: HeightMap = {
  micro: 32,
  small: 48,
  regular: 56,
  large: 64,
  jumbo: 108,
};

export const DEFAULT_WIDTH_PROPERTIES: WidthProperties = {
  width: 50,
  flexGrow: 1,
  flexShrink: 1,
  minWidth: 0,
};

export const EXPANDABLE_COLUMN_WIDTH = 50;

export const SELECTABLE_COLUMN_WIDTH = 50;
export const SELECTABLE_COLUMN_WIDTH_EXPANDABLE = 39;
