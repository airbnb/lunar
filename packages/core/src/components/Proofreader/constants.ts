import { ARROW_LEFT, ARROW_UP, ARROW_DOWN, ARROW_RIGHT } from '../../keys';

export const AIRBNB_REGEX = /\b(((air|ari|iar)[bn]{3})(?!\.com))\b/gi;
export const NON_WORD_REGEX = /\W/;

export const ARROW_KEYS = [ARROW_LEFT, ARROW_UP, ARROW_DOWN, ARROW_RIGHT];

export const NO_LOCALE = 'none';
export const AUTO_DETECT_LOCALE = 'auto';
export const LOCALE_TO_LT_LOCALE: { [locale: string]: string } = {
  de: 'de-DE',
  en: 'en-US',
  ja: 'ja-JP',
  pt: 'pt-PT',
  ru: 'ru-RU',
  zh: 'zh-CN',
};
