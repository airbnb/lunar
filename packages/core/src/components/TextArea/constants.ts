// eslint-disable-next-line import/prefer-default-export
export const LANGUAGE_RULES = {
  UNNECESSARY_ADMIN_NOTE_RULE: 'AGENT_TOOLS_UNNECESSARY_ADMIN_NOTE_RULE',
};
