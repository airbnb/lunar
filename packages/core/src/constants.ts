import { Currency, Locale, LocaleDefinition, TimeZone } from './types';

// I18N/L10N
export const DEFAULT_LOCALE: Locale = 'en';
export const DEFAULT_CURRENCY: Currency = 'USD';
export const DEFAULT_TIMEZONE: TimeZone = 'UTC';

export const LOCALES: Locale[] = [
  'de', // German
  'en', // English
  'es', // Spanish
  'fr', // French
  'it', // Italian
  'ja', // Japanese
  'ko', // Korean
  'ms', // Malay
  'nl', // Dutch
  'pt', // Portuguese
  'ru', // Russian
  'tr', // Turkish
  'zh', // Mandarin
];

// Used for Google Translate API
export const TRANSLATE_LOCALES: LocaleDefinition[] = [
  { locale: 'ar', label: 'ÿπÿ±ÿ®Ÿâ', name: 'Arabic' },
  { locale: 'ca', label: 'Catal√†', name: 'Catalan' },
  { locale: 'cs', label: 'ƒåe≈°tina', name: 'Czech' },
  { locale: 'da', label: 'Dansk', name: 'Danish' },
  { locale: 'de', label: 'Deutsch', name: 'German' },
  { locale: 'el', label: 'ŒïŒªŒªŒ∑ŒΩŒπŒ∫Œ¨', name: 'Greek' },
  { locale: 'en', label: 'English', name: 'English' },
  { locale: 'es', label: 'Espa√±ol', name: 'Spanish' },
  { locale: 'es-419', label: 'Espa√±ol (Am√©rica Latina)', name: 'Spanish (Latin America)' },
  { locale: 'fi', label: 'Suomalainen', name: 'Finnish' },
  { locale: 'fr', label: 'Fran√ßais', name: 'French' },
  { locale: 'he', label: '◊¢÷¥◊ë◊®÷¥◊ô◊™', name: 'Hebrew' },
  { locale: 'hr', label: 'Hrvatski', name: 'Croatian' },
  { locale: 'hu', label: 'Magyar', name: 'Hungarian' },
  { locale: 'id', label: 'Bahasa Indonesia', name: 'Indonesian' },
  { locale: 'is', label: '√çslensku', name: 'Icelandic' },
  { locale: 'it', label: 'Italiano', name: 'Italian' },
  { locale: 'ja', label: 'Êó•Êú¨Ë™û', name: 'Japanese' },
  { locale: 'ko', label: 'ÌïúÍµ≠Ïñ¥', name: 'Korean' },
  { locale: 'ms', label: 'Melayu', name: 'Malay' },
  { locale: 'nl', label: 'Nederlands', name: 'Dutch' },
  { locale: 'no', label: 'Norsk', name: 'Norwegian' },
  { locale: 'pl', label: 'Polskie', name: 'Polish' },
  { locale: 'pt', label: 'Portugu√™s', name: 'Portuguese' },
  { locale: 'pt-PT', label: 'Portugu√™s (Portugal)', name: 'Portuguese (Portugal)' },
  { locale: 'ru', label: '–†—É—Å—Å–∫–∏–π', name: 'Russian' },
  { locale: 'sv', label: 'Svenska', name: 'Swedish' },
  { locale: 'th', label: '‡πÑ‡∏ó‡∏¢', name: 'Thai' },
  { locale: 'tr', label: 'T√ºrk√ße', name: 'Turkish' },
  { locale: 'vi', label: 'Ti·∫øng Vi·ªát', name: 'Vietnamese' },
  { locale: 'zh', label: '‰∏≠Êñá (ÁÆÄ)', name: 'Chinese (Simplified)' },
  { locale: 'zh-TW', label: '‰∏≠Êñá (ÁπÅ)', name: 'Chinese (Traditional)' },
];

// Used for Language Tool (not all are listed)
// https://languagetool.org/http-api/swagger-ui/#!/default/get_languages
export const LT_LOCALES: LocaleDefinition[] = [
  { locale: 'ca-ES', label: 'Catal√†', name: 'Catalan' },
  { locale: 'ca-ES-valencia', label: 'Catal√† (Valenci√†)', name: 'Catalan' },
  { locale: 'da-DK', label: 'Dansk', name: 'Danish' },
  { locale: 'de-AT', label: 'Deutsch (√ñsterreich)', name: 'German (Austria)' },
  { locale: 'de-DE', label: 'Deutsch (Deutschland)', name: 'German (Germany)' },
  { locale: 'el', label: 'ŒïŒªŒªŒ∑ŒΩŒπŒ∫Œ¨', name: 'Greek' },
  { locale: 'en-AU', label: 'English (Australia)' },
  { locale: 'en-CA', label: 'English (Canada)' },
  { locale: 'en-GB', label: 'English (Great Britain)' },
  { locale: 'en-NZ', label: 'English (New Zealand)' },
  { locale: 'en-US', label: 'English (United States)' },
  { locale: 'es', label: 'Espa√±ol (Espa√±a)', name: 'Spanish (Spain)' },
  { locale: 'fr', label: 'Fran√ßais', name: 'French' },
  { locale: 'it', label: 'Italiano', name: 'Italian' },
  { locale: 'ja-JP', label: 'Êó•Êú¨Ë™û', name: 'Japanese' },
  // Korean ÌïúÍµ≠Ïñ¥
  // Malay Melayu
  { locale: 'nl', label: 'Nederlands', name: 'Dutch' },
  { locale: 'pl-PL', label: 'Polskie', name: 'Polish' },
  { locale: 'pt-PT', label: 'Portugu√™s (Portugal)', name: 'Portuguese (Portugal)' },
  { locale: 'pt-BR', label: 'Portugu√™s (Brasil)', name: 'Portuguese (Brazil)' },
  { locale: 'ro-RO', label: 'Rom√¢nƒÉ', name: 'Romanian' },
  { locale: 'ru-RU', label: '–†—É—Å—Å–∫–∏–π', name: 'Russian' },
  { locale: 'sv', label: 'Svenska', name: 'Swedish' },
  // Turkish T√ºrk√ße
  { locale: 'uk-UA', label: '–£–∫—Ä–∞—ó–Ω—Å—å–∫–∞', name: 'Ukrainian' },
  { locale: 'zh-CN', label: '‰∏≠Êñá (ÁÆÄ)', name: 'Chinese (Simplified)' },
];

// DATE/TIME
export const FORMAT_YMD = 'yyyy-MM-dd';
export const FORMAT_PREFIX_DAY = 'ccc'; // Sun
export const FORMAT_SUFFIX_TIMEZONE = 'ZZZZ'; // PST

// MISC
export const MICROS = 1000000;

// THEMES
export const STATUSES = ['notice', 'info', 'success', 'warning', 'danger', 'muted'];
export const BRANDS = ['luxury', 'plus'];
export const Z_INDEX_MODAL = 2000;
export const Z_INDEX_PORTAL = 2000; // Same as Modal to fix Tooltip in Modal
export const Z_INDEX_TOAST = 3000;

// EMOJIS
export const EMOJI_WHITELIST = [
  '1F600', // üòÄ grinning face
  '1F60A', // üòä smiling face with smiling eyes
  '1F642', // üôÇ slightly smiling face
  '1F389', // üéâ party popper
  '2728', // ‚ú® sparkles
  '1F334', // üå¥ palm tree
  '1F333', // üå≥ deciduous tree
  '1F335', // üåµ cactus
  '2600', // ‚òÄÔ∏è sun
  '2744', // ‚ùÑÔ∏è snowflake
  '1F3E0', // üè†Ô∏è house
  '1F697', // üöó automobile
  '2708', // ‚úàÔ∏è airplane
  '1F682', // üöÇ locomotive
  '26F5', // ‚õµÔ∏è sailboat
];

// REACT
export const PASSTHROUGH_WRAPPER_NAMES = ['Tooltip', 'Permission', 'Trebuchet', 'ERF', 'Memo'];
export const STRIP_HOC_NAMES = ['Memo', 'Proxy', 'withStyles', 'withTheme', 'connect'];
