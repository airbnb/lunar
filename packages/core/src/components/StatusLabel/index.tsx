import React from 'react';
import { mutuallyExclusiveTrueProps } from 'airbnb-prop-types';
import withStyles, { WithStylesProps } from '../../composers/withStyles';
import IconAffix from '../private/IconAffix';
import { STATUSES, BRANDS } from '../../constants';

const statusPropType = mutuallyExclusiveTrueProps(...STATUSES, ...BRANDS);

export type Props = {
  /** Icon to display at the end of the content. */
  afterIcon?: React.ReactNode;
  /** Icon to display at the start of the content. */
  beforeIcon?: React.ReactNode;
  /** Apply a border. */
  bordered?: boolean;
  /** Content within the label. */
  children: NonNullable<React.ReactNode>;
  /** Use compact padding. */
  compact?: boolean;
  /** Dangerous/failure status (red). */
  danger?: boolean;
  /** Informational status (blue). */
  info?: boolean;
  /** Invert background and text colors. */
  inverted?: boolean;
  /** Luxury brand. */
  luxury?: boolean;
  /** Muted/disabled status (gray). */
  muted?: boolean;
  /** Notice status. */
  notice?: boolean;
  /** Plus brand. */
  plus?: boolean;
  /** Successful status (green). */
  success?: boolean;
  /** Uppercase all text. */
  uppercased?: boolean;
  /** Warning status (yellow). */
  warning?: boolean;
};

/** Classify content through the use of tiny colorful status labels. */
export class StatusLabel extends React.Component<Props & WithStylesProps> {
  static propTypes = {
    danger: statusPropType,
    info: statusPropType,
    luxury: statusPropType,
    muted: statusPropType,
    notice: statusPropType,
    plus: statusPropType,
    success: statusPropType,
    warning: statusPropType,
  };

  static defaultProps = {
    afterIcon: null,
    beforeIcon: null,
    bordered: false,
    compact: false,
    danger: false,
    info: false,
    inverted: false,
    luxury: false,
    muted: false,
    notice: false,
    plus: false,
    success: false,
    uppercased: false,
    warning: false,
  };

  render() {
    const {
      cx,
      afterIcon,
      beforeIcon,
      bordered,
      children,
      compact,
      danger,
      info,
      inverted,
      luxury,
      muted,
      notice,
      plus,
      styles,
      success,
      uppercased,
      warning,
    } = this.props;

    return (
      <span
        className={cx(
          styles.label,
          uppercased && styles.label_uppercased,
          inverted && styles.label_inverted,
          bordered && styles.label_bordered,
          compact && styles.label_compact,
          danger && (inverted ? styles.label_inverted_danger : styles.label_danger),
          info && (inverted ? styles.label_inverted_info : styles.label_info),
          muted && (inverted ? styles.label_inverted_muted : styles.label_muted),
          notice && (inverted ? styles.label_inverted_notice : styles.label_notice),
          success && (inverted ? styles.label_inverted_success : styles.label_success),
          warning && (inverted ? styles.label_inverted_warning : styles.label_warning),
          luxury && (inverted ? styles.label_inverted_luxury : styles.label_luxury),
          plus && (inverted ? styles.label_inverted_plus : styles.label_plus),
        )}
      >
        {beforeIcon && <IconAffix before>{beforeIcon}</IconAffix>}

        <span>{children}</span>

        {afterIcon && <IconAffix after>{afterIcon}</IconAffix>}
      </span>
    );
  }
}

export default withStyles(({ color, font, ui, unit }) => ({
  label: {
    ...font.textMicro,
    display: 'inline-block',
    justifyContent: 'space-between',
    alignItems: 'center',
    verticalAlign: 'middle',
    whiteSpace: 'nowrap',
    marginRight: unit / 2,
    paddingLeft: unit,
    paddingRight: unit,
    paddingTop: unit / 2,
    paddingBottom: unit / 2,
    borderRadius: ui.borderRadius,
    backgroundColor: color.core.neutral[1],
    color: color.accent.text,
    fontWeight: font.weights.semibold,
    border: ui.border,
    borderColor: 'transparent',
  },

  label_uppercased: {
    textTransform: 'uppercase',
  },

  label_bordered: {
    borderColor: color.accent.border,
  },

  label_compact: {
    paddingLeft: 0.75 * unit,
    paddingRight: 0.75 * unit,
    paddingTop: unit / 4,
    paddingBottom: unit / 4,
  },

  label_inverted: {
    backgroundColor: color.clear,
    color: color.accent.text,
  },

  label_danger: {
    backgroundColor: color.core.danger[1],
  },

  label_inverted_danger: {
    color: color.core.danger[3],
  },

  label_info: {
    backgroundColor: color.core.primary[1],
  },

  label_inverted_info: {
    color: color.core.primary[3],
  },

  label_luxury: {
    backgroundColor: color.brand.luxury[3],
    color: color.base,
  },

  label_inverted_luxury: {
    color: color.brand.luxury[3],
  },

  label_muted: {
    backgroundColor: color.core.neutral[2],
  },

  label_inverted_muted: {
    color: color.core.neutral[3],
  },

  label_notice: {
    backgroundColor: color.core.secondary[1],
  },

  label_inverted_notice: {
    color: color.core.secondary[3],
  },

  label_plus: {
    backgroundColor: color.brand.plus[3],
    color: color.base,
  },

  label_inverted_plus: {
    color: color.brand.plus[3],
  },

  label_success: {
    backgroundColor: color.core.success[1],
  },

  label_inverted_success: {
    color: color.core.success[3],
  },

  label_warning: {
    backgroundColor: color.core.warning[1],
  },

  label_inverted_warning: {
    color: color.core.warning[6],
  },
}))(StatusLabel);
