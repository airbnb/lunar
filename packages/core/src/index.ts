import Aesthetic, { FontFace } from 'aesthetic';
import AphroditeAesthetic, { NativeBlock, ParsedBlock } from 'aesthetic-adapter-aphrodite';
import { Settings as LuxonSettings } from 'luxon';
import { Path as EmojiPath } from 'interweave-emoji';
import globalStyles from './themes/global';
import lightTheme from './themes/light';
import darkTheme from './themes/dark';
import getLocaleFromClient from './utils/getLocaleFromClient';
import getTimezoneFromClient from './utils/getTimezoneFromClient';
import { DEFAULT_LOCALE, DEFAULT_TIMEZONE } from './constants';
import {
  Locale,
  TimeZone,
  Logger,
  Translator,
  TranslateParams,
  TranslateProps,
  Theme,
} from './types';

export type Settings = {
  defaultLocale?: Locale;
  defaultTimezone?: TimeZone;
  emojiCDN?: EmojiPath;
  errorURL?: string;
  fontFaces?: { [fontFamily: string]: FontFace[] };
  fontFamily?: string;
  logger?: Logger | null;
  name: string;
  theme?: 'light' | 'dark';
  translator?: Translator | null;
  translatorComponent?: React.ComponentType<TranslateProps> | null;
};

class Core {
  settings: Required<Settings> = {
    defaultLocale: getLocaleFromClient() || DEFAULT_LOCALE,
    defaultTimezone: getTimezoneFromClient() || DEFAULT_TIMEZONE,
    emojiCDN: '',
    errorURL: '',
    fontFaces: {},
    fontFamily:
      '-apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol"',
    logger: null,
    name: 'Lunar',
    theme: 'light',
    translator: null,
    translatorComponent: null,
  };

  protected aesthetic?: Aesthetic<Theme, NativeBlock, ParsedBlock>;

  initialize(settings: Settings) {
    this.settings = {
      ...this.settings,
      ...settings,
    };

    this.bootstrapAesthetic();
    this.bootstrapLuxon();
  }

  bootstrapAesthetic() {
    const { theme, fontFaces } = this.settings;
    const fontFamily = this.fontFamily();
    const globals = globalStyles(fontFaces);

    this.aesthetic = new AphroditeAesthetic<Theme>([], {
      theme,
      passThemeProp: false,
      pure: true,
    })
      .registerTheme('light', lightTheme(fontFamily), globals)
      .registerTheme('dark', darkTheme(fontFamily), globals);
  }

  bootstrapLuxon() {
    LuxonSettings.defaultLocale = this.locale();
    LuxonSettings.defaultZoneName = this.timezone();
    LuxonSettings.throwOnInvalid = true;
  }

  getAesthetic() {
    if (__DEV__) {
      if (!this.aesthetic) {
        throw new Error(
          'Aesthetic has not been initialized. Please call `Core.initialize()` from `@airbnb/lunar`.',
        );
      }
    }

    return this.aesthetic!;
  }

  fontFamily(): string {
    const locale = this.locale();

    if (locale.startsWith('ja')) {
      return '"ヒラギノ角ゴ Pro", "Hiragino Kaku Gothic Pro", メイリオ, Meiryo, Osaka, "ＭＳ Ｐゴシック", "MS PGothic", "MS Gothic", "ＭＳ ゴシック", "Helvetica Neue", Helvetica, Arial, sans-serif';
    }

    if (locale.startsWith('ko')) {
      return '"나눔 고딕", "Nanum Gothic", "맑은 고딕", "Malgun Gothic", "Apple Gothic", 돋움, Dotum, "Helvetica Neue", Helvetica, Arial, sans-serif';
    }

    if (locale.includes('zh')) {
      return '"Hiragino Sans GB", 华文细黑, STHeiti, 微软雅黑, "Microsoft YaHei", SimHei, "Helvetica Neue", Helvetica, Arial, sans-serif';
    }

    return this.settings.fontFamily;
  }

  locale(): Locale {
    return this.settings.defaultLocale || DEFAULT_LOCALE;
  }

  log = (error: Error, extra: object) => {
    const { logger } = this.settings;

    if (logger) {
      logger(error, extra);
    }
  };

  timezone(): TimeZone {
    return this.settings.defaultTimezone || DEFAULT_TIMEZONE;
  }

  translate = (phrase: string, params: TranslateParams, context: string) => {
    const { translator } = this.settings;

    if (translator) {
      return translator(phrase, params, context);
    }

    let message = phrase;

    if (phrase.includes('||||')) {
      const [singular, plural] = phrase.split('||||');
      const count = params.smartCount || 0;

      message = count === 1 ? singular : plural;
    }

    // Low-level token interpolation
    return message.replace(/%\{(\w+)\}/g, (match, key) => String(params[key] || ''));
  };
}

export default new Core();
