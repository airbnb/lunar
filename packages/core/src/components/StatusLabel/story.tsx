import React from 'react';
import IconAddAlt from '@airbnb/lunar-icons/lib/interface/IconAddAlt';
import Text from '../Text';
import StatusLabel from '.';

export default {
  title: 'Core/StatusLabel',
  parameters: {
    inspectComponents: [StatusLabel],
  },
};

export function defaultAndStatusLabels() {
  return (
    <>
      <StatusLabel>Default</StatusLabel>
      <StatusLabel notice>Notice</StatusLabel>
      <StatusLabel info>Info</StatusLabel>
      <StatusLabel success>Success</StatusLabel>
      <StatusLabel warning>Warning</StatusLabel>
      <StatusLabel danger>Danger</StatusLabel>
      <StatusLabel muted>Muted</StatusLabel>
    </>
  );
}

defaultAndStatusLabels.story = {
  name: 'Default and status labels.',
};

export function brandedStatusLabels() {
  return (
    <>
      <StatusLabel luxury>Luxury</StatusLabel>
      <StatusLabel plus>Plus</StatusLabel>
    </>
  );
}

brandedStatusLabels.story = {
  name: 'Branded status labels.',
};

export function colorInvertedAndUppercasedLabels() {
  return (
    <>
      <StatusLabel inverted uppercased>
        Default
      </StatusLabel>
      <StatusLabel inverted uppercased notice>
        Notice
      </StatusLabel>
      <StatusLabel inverted uppercased info>
        Info
      </StatusLabel>
      <StatusLabel inverted uppercased success>
        Success
      </StatusLabel>
      <StatusLabel inverted uppercased warning>
        Warning
      </StatusLabel>
      <StatusLabel inverted uppercased danger>
        Danger
      </StatusLabel>
      <StatusLabel inverted uppercased muted>
        Muted
      </StatusLabel>
      <StatusLabel inverted uppercased luxury>
        Luxury
      </StatusLabel>
      <StatusLabel inverted uppercased plus>
        Plus
      </StatusLabel>
    </>
  );
}

colorInvertedAndUppercasedLabels.story = {
  name: 'Color inverted and uppercased labels.',
};

export function withABorderApplied() {
  return (
    <StatusLabel inverted bordered>
      Default
    </StatusLabel>
  );
}

withABorderApplied.story = {
  name: 'With a border applied.',
};

export function withinABlockOfText() {
  return (
    <Text>
      Content before the label. <StatusLabel>Default</StatusLabel> Content after the label.
    </Text>
  );
}

withinABlockOfText.story = {
  name: 'Within a block of text.',
};

export function withBeforeAndOrAfterIcons() {
  return (
    <>
      <div>
        <StatusLabel success beforeIcon={<IconAddAlt decorative />}>
          Before icon
        </StatusLabel>
        <StatusLabel notice afterIcon={<IconAddAlt decorative />}>
          After icon
        </StatusLabel>
      </div>
    </>
  );
}

withBeforeAndOrAfterIcons.story = {
  name: 'With before and or after icons.',
};

export function withCompactPadding() {
  return (
    <>
      <StatusLabel compact uppercased>
        Compact
      </StatusLabel>
      <StatusLabel info compact uppercased>
        Info
      </StatusLabel>
      <StatusLabel compact info inverted bordered uppercased>
        Info Inverted
      </StatusLabel>
    </>
  );
}

withCompactPadding.story = {
  name: 'With compact padding.',
};

export function sameHeightWithOrWithoutBorderApplied() {
  return (
    <>
      <StatusLabel>Default</StatusLabel>
      <StatusLabel inverted bordered>
        Default
      </StatusLabel>
    </>
  );
}

sameHeightWithOrWithoutBorderApplied.story = {
  name: 'Same height with or without border applied',
};
