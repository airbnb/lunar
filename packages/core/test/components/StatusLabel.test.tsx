import React from 'react';
import { shallowWithStyles } from '@airbnb/lunar-test-utils';
import IconCheck from '@airbnb/lunar-icons/lib/interface/IconCheck';
import StatusLabel from '../../src/components/StatusLabel';
import { STATUSES as BASE_STATUSES } from '../../src/constants';

const STATUSES = [...BASE_STATUSES, 'luxury', 'plus'];

describe('<StatusLabel />', () => {
  it('errors when multiple states are used at once', () => {
    expect(() => {
      shallowWithStyles(
        <StatusLabel danger success>
          Default
        </StatusLabel>,
      );
    }).toThrowError();
  });

  it('renders a before icon', () => {
    const icon = <IconCheck decorative />;
    const wrapper = shallowWithStyles(<StatusLabel beforeIcon={icon}>Default</StatusLabel>);

    expect(wrapper.contains(icon)).toBe(true);
  });

  it('renders a after icon', () => {
    const icon = <IconCheck decorative />;
    const wrapper = shallowWithStyles(<StatusLabel afterIcon={icon}>Default</StatusLabel>);

    expect(wrapper.contains(icon)).toBe(true);
  });

  it('renders both icons', () => {
    const beforeIcon = <IconCheck decorative />;
    const afterIcon = <IconCheck decorative />;
    const wrapper = shallowWithStyles(
      <StatusLabel beforeIcon={beforeIcon} afterIcon={afterIcon}>
        Default
      </StatusLabel>,
    );

    expect(wrapper.contains(beforeIcon)).toBe(true);
    expect(wrapper.contains(afterIcon)).toBe(true);
  });

  describe('statuses', () => {
    STATUSES.forEach(status => {
      it('renders label', () => {
        const wrapper = shallowWithStyles(
          <StatusLabel {...{ [status]: true }}>{status}</StatusLabel>,
        );

        expect(wrapper).toMatchSnapshot();
      });

      it('renders bordered', () => {
        const wrapper = shallowWithStyles(
          <StatusLabel bordered {...{ [status]: true }}>
            {status}
          </StatusLabel>,
        );

        expect(wrapper).toMatchSnapshot();
      });

      it('renders inverted state', () => {
        const wrapper = shallowWithStyles(
          <StatusLabel inverted {...{ [status]: true }}>
            {status}
          </StatusLabel>,
        );

        expect(wrapper).toMatchSnapshot();
      });

      it('renders uppercased state', () => {
        const wrapper = shallowWithStyles(
          <StatusLabel uppercased {...{ [status]: true }}>
            {status}
          </StatusLabel>,
        );

        expect(wrapper).toMatchSnapshot();
      });
    });
  });
});
