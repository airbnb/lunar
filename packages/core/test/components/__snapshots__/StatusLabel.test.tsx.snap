// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StatusLabel /> statuses renders bordered 1`] = `
<span
  className="label_xffcmo-o_O-label_bordered_1f3aaih-o_O-label_notice_h3lmxn"
>
  <span>
    notice
  </span>
</span>
`;

exports[`<StatusLabel /> statuses renders bordered 2`] = `
<span
  className="label_xffcmo-o_O-label_bordered_1f3aaih-o_O-label_info_5ni0mj"
>
  <span>
    info
  </span>
</span>
`;

exports[`<StatusLabel /> statuses renders bordered 3`] = `
<span
  className="label_xffcmo-o_O-label_bordered_1f3aaih-o_O-label_success_9ngehh"
>
  <span>
    success
  </span>
</span>
`;

exports[`<StatusLabel /> statuses renders bordered 4`] = `
<span
  className="label_xffcmo-o_O-label_bordered_1f3aaih-o_O-label_warning_kas06q"
>
  <span>
    warning
  </span>
</span>
`;

exports[`<StatusLabel /> statuses renders bordered 5`] = `
<span
  className="label_xffcmo-o_O-label_bordered_1f3aaih-o_O-label_danger_p8x1nu"
>
  <span>
    danger
  </span>
</span>
`;

exports[`<StatusLabel /> statuses renders bordered 6`] = `
<span
  className="label_xffcmo-o_O-label_bordered_1f3aaih-o_O-label_muted_s1bxkr"
>
  <span>
    muted
  </span>
</span>
`;

exports[`<StatusLabel /> statuses renders bordered 7`] = `
<span
  className="label_xffcmo-o_O-label_bordered_1f3aaih-o_O-label_luxury_1ue0q2o"
>
  <span>
    luxury
  </span>
</span>
`;

exports[`<StatusLabel /> statuses renders bordered 8`] = `
<span
  className="label_xffcmo-o_O-label_bordered_1f3aaih-o_O-label_plus_ncs1rt"
>
  <span>
    plus
  </span>
</span>
`;

exports[`<StatusLabel /> statuses renders inverted state 1`] = `
<span
  className="label_xffcmo-o_O-label_inverted_1jg2760-o_O-label_inverted_notice_a0aok"
>
  <span>
    notice
  </span>
</span>
`;

exports[`<StatusLabel /> statuses renders inverted state 2`] = `
<span
  className="label_xffcmo-o_O-label_inverted_1jg2760-o_O-label_inverted_info_13z9huf"
>
  <span>
    info
  </span>
</span>
`;

exports[`<StatusLabel /> statuses renders inverted state 3`] = `
<span
  className="label_xffcmo-o_O-label_inverted_1jg2760-o_O-label_inverted_success_1dekfjx"
>
  <span>
    success
  </span>
</span>
`;

exports[`<StatusLabel /> statuses renders inverted state 4`] = `
<span
  className="label_xffcmo-o_O-label_inverted_1jg2760-o_O-label_inverted_warning_sew9to"
>
  <span>
    warning
  </span>
</span>
`;

exports[`<StatusLabel /> statuses renders inverted state 5`] = `
<span
  className="label_xffcmo-o_O-label_inverted_1jg2760-o_O-label_inverted_danger_15o6i4l"
>
  <span>
    danger
  </span>
</span>
`;

exports[`<StatusLabel /> statuses renders inverted state 6`] = `
<span
  className="label_xffcmo-o_O-label_inverted_1jg2760-o_O-label_inverted_muted_q8d046"
>
  <span>
    muted
  </span>
</span>
`;

exports[`<StatusLabel /> statuses renders inverted state 7`] = `
<span
  className="label_xffcmo-o_O-label_inverted_1jg2760-o_O-label_inverted_luxury_1tvijvk"
>
  <span>
    luxury
  </span>
</span>
`;

exports[`<StatusLabel /> statuses renders inverted state 8`] = `
<span
  className="label_xffcmo-o_O-label_inverted_1jg2760-o_O-label_inverted_plus_1l928rd"
>
  <span>
    plus
  </span>
</span>
`;

exports[`<StatusLabel /> statuses renders label 1`] = `
<span
  className="label_xffcmo-o_O-label_notice_h3lmxn"
>
  <span>
    notice
  </span>
</span>
`;

exports[`<StatusLabel /> statuses renders label 2`] = `
<span
  className="label_xffcmo-o_O-label_info_5ni0mj"
>
  <span>
    info
  </span>
</span>
`;

exports[`<StatusLabel /> statuses renders label 3`] = `
<span
  className="label_xffcmo-o_O-label_success_9ngehh"
>
  <span>
    success
  </span>
</span>
`;

exports[`<StatusLabel /> statuses renders label 4`] = `
<span
  className="label_xffcmo-o_O-label_warning_kas06q"
>
  <span>
    warning
  </span>
</span>
`;

exports[`<StatusLabel /> statuses renders label 5`] = `
<span
  className="label_xffcmo-o_O-label_danger_p8x1nu"
>
  <span>
    danger
  </span>
</span>
`;

exports[`<StatusLabel /> statuses renders label 6`] = `
<span
  className="label_xffcmo-o_O-label_muted_s1bxkr"
>
  <span>
    muted
  </span>
</span>
`;

exports[`<StatusLabel /> statuses renders label 7`] = `
<span
  className="label_xffcmo-o_O-label_luxury_1ue0q2o"
>
  <span>
    luxury
  </span>
</span>
`;

exports[`<StatusLabel /> statuses renders label 8`] = `
<span
  className="label_xffcmo-o_O-label_plus_ncs1rt"
>
  <span>
    plus
  </span>
</span>
`;

exports[`<StatusLabel /> statuses renders uppercased state 1`] = `
<span
  className="label_xffcmo-o_O-label_uppercased_4efw5a-o_O-label_notice_h3lmxn"
>
  <span>
    notice
  </span>
</span>
`;

exports[`<StatusLabel /> statuses renders uppercased state 2`] = `
<span
  className="label_xffcmo-o_O-label_uppercased_4efw5a-o_O-label_info_5ni0mj"
>
  <span>
    info
  </span>
</span>
`;

exports[`<StatusLabel /> statuses renders uppercased state 3`] = `
<span
  className="label_xffcmo-o_O-label_uppercased_4efw5a-o_O-label_success_9ngehh"
>
  <span>
    success
  </span>
</span>
`;

exports[`<StatusLabel /> statuses renders uppercased state 4`] = `
<span
  className="label_xffcmo-o_O-label_uppercased_4efw5a-o_O-label_warning_kas06q"
>
  <span>
    warning
  </span>
</span>
`;

exports[`<StatusLabel /> statuses renders uppercased state 5`] = `
<span
  className="label_xffcmo-o_O-label_uppercased_4efw5a-o_O-label_danger_p8x1nu"
>
  <span>
    danger
  </span>
</span>
`;

exports[`<StatusLabel /> statuses renders uppercased state 6`] = `
<span
  className="label_xffcmo-o_O-label_uppercased_4efw5a-o_O-label_muted_s1bxkr"
>
  <span>
    muted
  </span>
</span>
`;

exports[`<StatusLabel /> statuses renders uppercased state 7`] = `
<span
  className="label_xffcmo-o_O-label_uppercased_4efw5a-o_O-label_luxury_1ue0q2o"
>
  <span>
    luxury
  </span>
</span>
`;

exports[`<StatusLabel /> statuses renders uppercased state 8`] = `
<span
  className="label_xffcmo-o_O-label_uppercased_4efw5a-o_O-label_plus_ncs1rt"
>
  <span>
    plus
  </span>
</span>
`;
